src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -4755,7 +4810,9 @@ public static bool ConnectAsync(SocketType socketType, ProtocolType protocolType
                 MultipleConnectAsync multipleConnectAsync = null;
                 if (dnsEP.AddressFamily == AddressFamily.Unspecified)
                 {
-                    multipleConnectAsync = new MultipleSocketMultipleConnectAsync(socketType, protocolType);
+                    multipleConnectAsync = SocketPal.SupportsMultipleConnectAttempts ?
+                        (MultipleConnectAsync)(new DualSocketMultipleConnectAsync(socketType, protocolType)) :
```

Comment:
> Why are we getting so much code churn to support OSX instead of pushing this code in PAL?

It's not just churn to support OS X specifically: it's churn to enable *any* platform that does not support multiple connect attempts on the same socket (e.g. some/all of the BSDs).

> I think at this point we'd need some sort of architecture design that explains these changes. The CR became too complicated to understand the intent.

I'll split these changes out into their own PR.> Why are we getting so much code churn to support OSX instead of pushing this code in PAL?

It's not just churn to support OS X specifically: it's churn to enable *any* platform that does not support multiple connect attempts on the same socket (e.g. some/all of the BSDs).

> I think at this point we'd need some sort of architecture design that explains these changes. The CR became too complicated to understand the intent.

I'll split these changes out into their own PR.